Add scan page splitting and empty regions.

marijnkoolen opened the pull request:
Improvements:
- Add function `pagexml.analysis.stats.get_doc_region_stats` to get statistics for regions below scan-level.
- Add function `pagexml.parsers.scan_parser.split_scan_pages_with_separation_point` to split a scan instance into two page instances using a separation point.
- Add functions in `pagexml.helper.spatial_helper` to check whether two regions touch each other without overlapping.
- Add generic `PageXMLRegion` base class to the PageXML Document Model and make `PageXMLTextRegion`, `PageXMLTableRegion` and `PageXMLEmptyRegion` inherit from this base class. This simplifies selecting either all regions or only regions of a specific type.
- Add class `PageXMLEmptyRegion` to generate empty regions (to make explicit which regions of a scan or page are empty).
- Improve copying PageXMLDoc objects to avoid copying parents. This makes copying much faster and avoids generating stray objects.
- Improve merging of columns via `pagexml.parsers.column_parser.merge_columns` (including overlapping columns via `column_parser.normalise_columns`) to properly carry over metadata and group text lines in a `PageXMLTextRegion` object, so that the columns no longer have `PageXMLTextLine`s as direct children.

Bug fixes:
- Fix bug generating Coords for points that have either zero width or zero height.
- Fix bug in `set_parentage` that skipped new region types.
- Fix bug in `pagexml.parser` functions `json_to_pagexml_column`, `json_to_pagexml_page` and `json_to_pagexml_scan` that ignored PageXMLTableRegion elements.

marijnkoolen committed:
Add scan page splitting and empty regions.

Improvements:
- Add function `pagexml.analysis.stats.get_doc_region_stats` to get statistics for regions below scan-level.
- Add function `pagexml.parsers.scan_parser.split_scan_pages_with_separation_point` to split a scan instance into two page instances using a separation point.
- Add functions in `pagexml.helper.spatial_helper` to check whether two regions touch each other without overlapping.
- Add generic `PageXMLRegion` base class to the PageXML Document Model and make `PageXMLTextRegion`, `PageXMLTableRegion` and `PageXMLEmptyRegion` inherit from this base class. This simplifies selecting either all regions or only regions of a specific type.
- Add class `PageXMLEmptyRegion` to generate empty regions (to make explicit which regions of a scan or page are empty).
- Improve copying PageXMLDoc objects to avoid copying parents. This makes copying much faster and avoids generating stray objects.
- Improve merging of columns via `pagexml.parsers.column_parser.merge_columns` (including overlapping columns via `column_parser.normalise_columns`) to properly carry over metadata and group text lines in a `PageXMLTextRegion` object, so that the columns no longer have `PageXMLTextLine`s as direct children.

Bug fixes:
- Fix bug generating Coords for points that have either zero width or zero height.
- Fix bug in `set_parentage` that skipped new region types.
- Fix bug in `pagexml.parser` functions `json_to_pagexml_column`, `json_to_pagexml_page` and `json_to_pagexml_scan` that ignored PageXMLTableRegion elements.

marijnkoolen committed:
Fix bugs parsing custom attributes.

Improvements:
- Add `text_height` property that returns estimated height of the text line, which is the same as `xheight` if that is set, otherwise, it is estimated as 50% of the bounding box height (assuming 25% above and below the average character height).
- Improve `get_vertical_overlap` in `pagexml.model.pagexml_document_model` to use the `baseline` property if it is set to determine vertical overlap.

Bug fixes:
- Fix bug in `PageXMLScan.to_pagexml` returning an empty Page element (`_to_pagexml` was not implemented).
- Fix bug in parsing of custom attributes leading to an error because of properties that have no value (e.g. `strikethrough`). Now such properties are set as keys in the custom attribute dictionary with the value True, and in the conversion back to XML are properly set.
- Fix bug in generating the custom attribute string when converting PageXMLDoc objects to XML (the semicolon separator was not used).

marijnkoolen committed:
Fix bug in `has_baseline`

- Remove redudant and bugy version of `has_baseline` from
  `pagexml.model.physical_document_model` and import better version from
  `pagexml.model.pagexml_document_model
  `

marijnkoolen committed:
Improvements:

- calculation of `vertical_distance` now uses baslines of
  PageXMLTextLines if they are present, otherwise falls back to coords.
- determining whether two lines are next to each other in
  `PageXMLTextLine.is_next_to` now uses similar logic as `is_below`.

marijnkoolen committed:
Bug fixes:
- The `.regions` property of `PageXMLPage` objects didn't return any
  regions from columns and extra properties. Now all regions are
  returned.

brambg commented:
The tests fail because

> 'PageXMLPage' object has no attribute 'columns'

https://github.com/knaw-huc/pagexml/actions/runs/20774732164/job/59657775833?pr=22#step:7:213

Was the removing of the `columns` attribute intentional? What replaces this functionality?

marijnkoolen committed:
Fix PageXMLPage regions bug.

Improvements:
- Use distance between baselines for computing the vertical distance between two `PageXMLTextLine`s.

Bug fixes:
- Ensure that a line directly below a previously horizontally grouped line as assigned to a new group.
- Ensure `PageXMLRegion.regions` is not called in __init__ to avoid AttributeError (`'PageXMLPage' object has no attribute 'columns'`) when generating `PageXMLPage` object.

marijnkoolen committed:
Bug fixes:
- Fix line counting in PageXMLPage.stats.

marijnkoolen commented:
The `columns` property was not removed. The error was triggered before the `columns` property was set. Now fixed.

marijnkoolen committed:
Fix double group assignment bug
